include/aws/common/string.h

```diff
@@ -0,0 +1,101 @@
+#ifndef AWS_COMMON_STRING_H
+#define AWS_COMMON_STRING_H
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+#include <aws/common/common.h>
+#include <aws/common/byte_buf.h>
+#include <string.h>
+
+/**
+ * Represents an immutable string holding either text or binary data. If the string is in constant
+ * memory or memory that should otherwise not be freed by this struct, set allocator to NULL
+ * and destroy function will be a no-op.
+ *
+ * This is for use cases where the entire struct and the data bytes themselves need to be
+ * held in dynamic memory, such as when held by a struct aws_hash_table. The data bytes
+ * themselves are always held in contiguous memory immediately after the end of the
+ * struct aws_string, and the memory for both the header and the data bytes
+ * is allocated together.
+ *
+ * Use the aws_string_bytes function to access the data bytes.
+ */
+struct aws_string {
+    struct aws_allocator * allocator;
+    size_t len;
```

Comment:
Why not just put a uint8_t* at the end of this instead of the point arithmetic?

Reply:
This is used in a situation where we need to run malloc (or aws_mem_acquire) on the struct itself. If we allocated memory to a pointer within the struct, we would have to do two allocations, one for the struct and one for the pointer. The way it is coded now is an optimization to require only one malloc for the entire string (header struct and data bytes).

I suppose it would be possible to include a pointer to the data bytes and always set that pointer to the value of aws_string_bytes() upon creation. That would have the advantage of making the implementation opaque to the users and more like the byte buffer object, though it seems like an unnecessary use of memory, since the pointer will always do the same thing as the aws_string_bytes() function. (Not only that, but I would argue that using the pointer to always go the same number of bytes forward rather than the function is less safe from a security perspective.)

@bdonlan and I had a detailed conversation about this, and one insight he had was that in cases where we wanted to be able to reallocate the data bytes, it makes sense to allocate the struct and the data bytes separately, so that we can reallocate the memory within the string (such as when appending bytes beyond the current capacity) without losing the pointer to the original struct, which can be important when that struct is held in other container objects by that pointer. The optimization we made of allocating them all together makes sense in the context of an immutable string, and this use case of an immutable string is what is most useful to us in a hash table.